include/dali/core/exec/tasking/task.h

```diff
@@ -0,0 +1,522 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_EXEC_TASKING_TASK_H_
+#define DALI_CORE_EXEC_TASKING_TASK_H_
+
+#include <any>
+#include <cassert>
+#include <condition_variable>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+#include <tuple>
+#include <utility>
+
+#include "dali/core/exec/tasking/sync.h"
+#include "dali/core/small_vector.h"
+
+namespace dali::tasking {
+
+namespace detail {
+class TaskList;
+
+template <typename T, typename U = void>
+struct is_iterable : std::false_type {};
+
+template <typename T>
+struct is_iterable<T, std::void_t<
+  decltype(*std::begin(std::declval<T>())),
+  decltype(*std::end(std::declval<T>()))>>
+: std::true_type {};
+
+template <typename T>
+constexpr bool is_iterable_v = is_iterable<T>::value;
+
+template <typename T>
+struct is_tuple : std::false_type {};
+
+template <typename... Args>
+struct is_tuple<std::tuple<Args...>> : std::true_type {};
+
+template <typename T>
+constexpr bool is_tuple_v = is_tuple<T>::value;
+
+}  // namespace detail
+
+class Scheduler;
+
+
+class TaskResult {
+ public:
+  template <typename F>
+  void SetResultOf(F &&f) {
+    try {
+      if constexpr (std::is_void_v<decltype(f())>) {
+        f();
+        value_ = void_result();
+      } else {
+        value_ = f();
+      }
+    } catch (...) {
+      exception_ = std::current_exception();
+    }
+  }
+
+  template <typename T>
+  void Set(T &&t) {
+    value_ = std::forward<T>(t);
+  }
+
+  void SetException(std::exception_ptr e) {
+    exception_ = std::move(e);
+  }
+
+  bool Empty() {
+    return !value_.has_value();
+  }
+
+  const std::any &Value() const {
+    if (exception_)
+      std::rethrow_exception(exception_);
+    return value_;
+  }
+
+  template <typename T>
+  T Value() const {
+    if (exception_)
+      std::rethrow_exception(exception_);
+    if constexpr (std::is_void_v<T>) {
+      (void)std::any_cast<void_result>(value_);
+    }
+    return std::any_cast<T>(value_);
+  }
+
+  const bool HasException() const {
+    return exception_ != nullptr;
+  }
+
+ private:
+  struct void_result {};
+  std::any value_;
+  std::exception_ptr exception_;
+};
+
+using SharedTaskResult = std::shared_ptr<TaskResult>;
+
+static constexpr int ScalarResult = -1;
+
+class TaskResults : public SmallVector<SharedTaskResult, 4> {
+ public:
+  void Init(int num_results = ScalarResult) {
+    assert(num_results == ScalarResult || num_results > 0);
+    is_scalar_ = num_results < 0;
```

Comment:
Why not `== 1` ?
Negative value for scalar seems a bit surprising.

Reply:
There's a qualitative difference between 1 result and a scalar result - perhaps not here, but when the task is defined.
A task that has `num_results == 1` cannot return a single value - it must return a tuple or a collection.

Reply:
What is special in the scalar case here that it needs special handling? After all we need to index it the same way as it has more than one result.

Reply:
In TaskResults - nothing. I just made the handling uniform with Task, where it does differ.

Reply:
ok